src/Processors/Formats/Impl/FormInputFormat.cpp

```diff
@@ -0,0 +1,181 @@
+#include<Processors/Formats/Impl/FormInputFormat.h>
+#include "Formats/EscapingRuleUtils.h"
+#include <Formats/FormatFactory.h>
+
+namespace DB
+{
+
+namespace ErrorCodes
+{
+    extern const int INCORRECT_DATA;
+}
+
+FormInputFormat::FormInputFormat(ReadBuffer & in_, Block header_, Params params_, const FormatSettings & format_settings_) : IRowInputFormat(std::move(header_), in_, params_), format_settings(format_settings_)
+{
+    const auto & header = getPort().getHeader();
+    size_t num_columns = header.columns();
+    for (size_t i = 0; i < num_columns; ++i)
+        name_map[header.getByPosition(i).name] = i;
+}
+
+void FormInputFormat::readPrefix()
+{
+    skipBOMIfExists(*in);
+}
+
+const String & FormInputFormat::columnName(size_t i) const
+{
+    return getPort().getHeader().getByPosition(i).name;
+}
+
+void FormInputFormat::readField(size_t index, MutableColumns & columns)
+{
+    if (seen_columns[index])
+        throw Exception(ErrorCodes::INCORRECT_DATA, "Duplicate field found while parsing Form format: {}", columnName(index));
+
+    seen_columns[index] = read_columns[index] = true;
+    const auto & serialization = serializations[index];
+
+    String encoded_str, decoded_str;
+    readStringUntilAmpersand(encoded_str,*in);
+
+    if (!in->eof())
+        assertChar('&',*in);
+
+    Poco::URI::decode(encoded_str, decoded_str);
+    ReadBufferFromString buf(decoded_str);
+    serialization->deserializeWholeText(*columns[index], buf, format_settings);
+}
+
+String readFieldName(ReadBuffer & buf)
+{
+    String field;
+    readStringUntilEquals(field, buf);
+    assertChar('=', buf);
+    return field;
+}
+
+void FormInputFormat::readFormData(MutableColumns & columns)
+{
+    size_t index = 0;
+    StringRef name_ref;
+    while (true)
+    {
+        if (in->eof())
+            break;
+
+        auto tmp = readFieldName(*in);
+        name_ref = StringRef(tmp);
+        auto * it = name_map.find(name_ref);
+
+        if (!it)
+        {
+            if (!format_settings.skip_unknown_fields)
+                throw Exception(ErrorCodes::INCORRECT_DATA, "Unknown field found while parsing Form format: {}", name_ref.toString());
+
+            /// Skip the value if key is not found.
+            NullOutput sink;
+            String encoded_str;
+            readStringUntilAmpersand(encoded_str,*in);
+
+            if (!in->eof())
+                assertChar('&',*in);
+
+            ReadBufferFromString buf(encoded_str);
+            readStringInto(sink, buf);
```

Comment:
This is not needed, we already skipped this value in our `ReadBuffer`, no need to do anything else